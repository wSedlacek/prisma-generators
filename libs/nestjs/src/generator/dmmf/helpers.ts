// TODO: support not string args
export const modelAttributeRegex = /(@@NestJS\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const fieldAttributeRegex = /(@NestJS\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const attributeNameRegex = /(?:\.)+([A-Za-z])+(?:\()+/;
export const attributeArgsRegex = /(?:\(")+([A-Za-z])+(?:"\))+/;

export const parseDocumentationAttributes = (
  documentation: string | undefined,
  expectedAttributeName: string,
  expectedAttributeKind: 'model' | 'field'
) => {
  const attributeRegex =
    expectedAttributeKind === 'model'
      ? modelAttributeRegex
      : fieldAttributeRegex;
  const attribute = documentation?.match(attributeRegex)?.[0];
  const attributeName = attribute?.match(attributeNameRegex)?.[0]?.slice(1, -1);
  if (attributeName !== expectedAttributeName) {
    return undefined;
  }

  return attribute?.match(attributeArgsRegex)?.[0]?.slice(1, -1);
};
