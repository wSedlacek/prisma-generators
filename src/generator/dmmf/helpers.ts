// TODO: support not string args
export const modelAttributeRegex = /(@@TypeGraphQL\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const fieldAttributeRegex = /(@TypeGraphQL\.)+([A-z])+(\(")+([A-z])+("\))+/;
export const attributeNameRegex = /(?:\.)+([A-Za-z])+(?:\()+/;
export const attributeArgsRegex = /(?:\(")+([A-Za-z])+(?:"\))+/;

export function parseDocumentationAttributes(
  documentation: string | undefined,
  expectedAttributeName: string,
  expectedAttributeKind: "model" | "field",
) {
  const attributeRegex =
    expectedAttributeKind === "model"
      ? modelAttributeRegex
      : fieldAttributeRegex;
  const attribute = documentation?.match(attributeRegex)?.[0];
  const attributeName = attribute?.match(attributeNameRegex)?.[0]?.slice(1, -1);
  if (attributeName !== expectedAttributeName) {
    return;
  }
  const attributeArgs = attribute?.match(attributeArgsRegex)?.[0]?.slice(1, -1);
  return attributeArgs;
}
