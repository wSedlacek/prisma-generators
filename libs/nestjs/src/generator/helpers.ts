import { DmmfDocument } from './dmmf/dmmf-document';
import { fieldAttributeRegex, modelAttributeRegex } from './dmmf/helpers';
import { DMMF } from './dmmf/types';

export const noop = () => {};

export const getFieldTSType = (
  typeInfo: DMMF.TypeInfo,
  dmmfDocument: DmmfDocument,
  isInputType: boolean,
  modelName?: string,
  typeName?: string
) => {
  let TSType: string;
  if (typeInfo.kind === 'scalar') {
    TSType = mapScalarToTSType(typeInfo.type, isInputType);
  } else if (typeInfo.kind === 'object') {
    TSType =
      dmmfDocument.getModelTypeName(typeInfo.type) ?? (!typeName || !modelName)
        ? getInputTypeName(typeInfo.type, dmmfDocument)
        : typeInfo.type.replace(modelName, typeName);
  } else if (typeInfo.kind === 'enum') {
    TSType = `keyof typeof ${typeInfo.type}`;
  } else {
    throw new Error(`Unsupported field type kind: ${typeInfo.kind}`);
  }
  if (typeInfo.isList) {
    if (TSType.includes(' ')) {
      TSType = `Array<${TSType}>`;
    } else {
      TSType += '[]';
    }
  }
  if (!typeInfo.isRequired) {
    // FIXME: use properly null for output and undefined for input
    // TSType += ' | undefined';
    TSType += ' | null | undefined';
  }

  return TSType;
};

export const mapScalarToTSType = (scalar: string, isInputType: boolean) => {
  switch (scalar) {
    case 'ID':
    case 'UUID': {
      return 'string';
    }
    case 'String': {
      return 'string';
    }
    case 'Boolean': {
      return 'boolean';
    }
    case 'DateTime': {
      return 'Date';
    }
    case 'Int':
    case 'Float': {
      return 'number';
    }
    case 'Json':
      return isInputType ? 'InputJsonValue' : 'JsonValue';
    default:
      throw new Error(`Unrecognized scalar type: ${scalar}`);
  }
};

export const getGraphQLType = (
  typeInfo: DMMF.TypeInfo,
  dmmfDocument: DmmfDocument,
  modelName?: string,
  typeName?: string
) => {
  let GraphQLType: string;
  if (typeInfo.kind === 'scalar') {
    GraphQLType = mapScalarToType(typeInfo.type);
  } else if (typeInfo.kind === 'object') {
    GraphQLType =
      dmmfDocument.getModelTypeName(typeInfo.type) ?? (!typeName || !modelName)
        ? getInputTypeName(typeInfo.type, dmmfDocument)
        : typeInfo.type.replace(modelName, typeName);
  } else {
    GraphQLType = typeInfo.type;
  }
  if (typeInfo.isList) {
    GraphQLType = `[${GraphQLType}]`;
  }

  return GraphQLType;
};

export const mapScalarToType = (scalar: string) => {
  switch (scalar) {
    case 'DateTime': {
      return 'Date';
    }
    // TODO: use proper uuid graphql scalar
    case 'UUID': {
      return 'String';
    }
    case 'Boolean':
    case 'String': {
      return scalar;
    }
    case 'ID':
    case 'Int':
    case 'Float': {
      return `${scalar}`;
    }
    case 'Json': {
      return `GraphQLJSON`;
    }
    default: {
      throw new Error(`Unrecognized scalar type: ${scalar}`);
    }
  }
};

export const camelCase = (str: string) => {
  return str[0].toLowerCase() + str.slice(1);
};

export const pascalCase = (str: string): string => {
  return str[0].toUpperCase() + str.slice(1);
};

const getInputKeywordPhrasePosition = (inputTypeName: string) => {
  const inputParseResult = [
    'Create',
    'OrderBy',
    'Update',
    'Upsert',
    'ScalarWhere',
    'Where',
    'ListRelationFilter',
    'RelationFilter',
    'Filter',
  ]
    .map((inputKeyword) => inputTypeName.search(inputKeyword))
    .filter((position) => position >= 0);

  if (inputParseResult.length === 0) {
    return undefined;
  }

  return inputParseResult[0];
};

export const getModelNameFromInputType = (inputTypeName: string) => {
  const keywordPhrasePosition = getInputKeywordPhrasePosition(inputTypeName);
  if (!keywordPhrasePosition) {
    return undefined;
  }

  return inputTypeName.slice(0, keywordPhrasePosition);
};

export const getInputTypeName = (
  originalInputName: string,
  dmmfDocument: DmmfDocument
): string => {
  const keywordPhrasePosition = getInputKeywordPhrasePosition(
    originalInputName
  );
  if (!keywordPhrasePosition) {
    return originalInputName;
  }

  const modelName = originalInputName.slice(0, keywordPhrasePosition);
  const typeNameRest = originalInputName.slice(keywordPhrasePosition);
  const modelTypeName = dmmfDocument.getModelTypeName(modelName);
  if (!modelTypeName) {
    return originalInputName;
  }

  return `${modelTypeName}${typeNameRest}`;
};

export const cleanDocsString = (
  documentation: string | undefined
): string | undefined => {
  if (!documentation) {
    return undefined;
  }
  let cleanedDocs = documentation;
  cleanedDocs = cleanedDocs.replace(modelAttributeRegex, '');
  cleanedDocs = cleanedDocs.replace(fieldAttributeRegex, '');
  cleanedDocs = cleanedDocs.split('"').join('\\"');
  cleanedDocs = cleanedDocs.split('\r').join('');
  cleanedDocs = cleanedDocs.split('\\r').join('');
  cleanedDocs = cleanedDocs.split('\n').join('');
  cleanedDocs = cleanedDocs.split('\\n').join('');

  return cleanedDocs;
};

export const toUnixPath = (maybeWindowsPath: string) => {
  return maybeWindowsPath.split('\\').join('/');
};
