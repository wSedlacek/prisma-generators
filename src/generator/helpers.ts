import { DMMFTypeInfo } from "./types";
import { DmmfDocument } from "./dmmf/dmmf-document";
import { modelAttributeRegex, fieldAttributeRegex } from "./dmmf/helpers";

export function noop() {}

export function getFieldTSType(
  typeInfo: DMMFTypeInfo,
  dmmfDocument: DmmfDocument,
  isInputType: boolean,
  modelName?: string,
  typeName?: string,
) {
  let TSType: string;
  if (typeInfo.kind === "scalar") {
    TSType = mapScalarToTSType(typeInfo.type, isInputType);
  } else if (typeInfo.kind === "object") {
    if (dmmfDocument.isModelName(typeInfo.type)) {
      TSType = dmmfDocument.getModelTypeName(typeInfo.type)!;
    } else {
      TSType =
        !typeName || !modelName
          ? getInputTypeName(typeInfo.type, dmmfDocument)
          : typeInfo.type.replace(modelName, typeName);
    }
  } else if (typeInfo.kind === "enum") {
    TSType = `keyof typeof ${typeInfo.type}`;
  } else {
    throw new Error(`Unsupported field type kind: ${typeInfo.kind}`);
  }
  if (typeInfo.isList) {
    if (TSType.includes(" ")) {
      TSType = `Array<${TSType}>`;
    } else {
      TSType += "[]";
    }
  }
  if (!typeInfo.isRequired) {
    // FIXME: use properly null for output and undefined for input
    // TSType += " | null | undefined";
    TSType += " | undefined";
  }
  return TSType;
}

export function mapScalarToTSType(scalar: string, isInputType: boolean) {
  switch (scalar) {
    case "ID":
    case "UUID": {
      return "string";
    }
    case "String": {
      return "string";
    }
    case "Boolean": {
      return "boolean";
    }
    case "DateTime": {
      return "Date";
    }
    case "Int":
    case "Float": {
      return "number";
    }
    case "Json":
      return isInputType ? "InputJsonValue" : "JsonValue";
    default:
      throw new Error(`Unrecognized scalar type: ${scalar}`);
  }
}

export function getTypeGraphQLType(
  typeInfo: DMMFTypeInfo,
  dmmfDocument: DmmfDocument,
  modelName?: string,
  typeName?: string,
) {
  let GraphQLType: string;
  if (typeInfo.kind === "scalar") {
    GraphQLType = mapScalarToTypeGraphQLType(typeInfo.type);
  } else if (typeInfo.kind === "object") {
    if (dmmfDocument.isModelName(typeInfo.type)) {
      GraphQLType = dmmfDocument.getModelTypeName(typeInfo.type)!;
    } else {
      GraphQLType =
        !typeName || !modelName
          ? getInputTypeName(typeInfo.type, dmmfDocument)
          : typeInfo.type.replace(modelName, typeName);
    }
  } else {
    GraphQLType = typeInfo.type;
  }
  if (typeInfo.isList) {
    GraphQLType = `[${GraphQLType}]`;
  }
  return GraphQLType;
}

export function mapScalarToTypeGraphQLType(scalar: string) {
  switch (scalar) {
    case "DateTime": {
      return "Date";
    }
    // TODO: use proper uuid graphql scalar
    case "UUID": {
      return "String";
    }
    case "Boolean":
    case "String": {
      return scalar;
    }
    case "ID":
    case "Int":
    case "Float": {
      return `${scalar}`;
    }
    case "Json": {
      return `GraphQLJSON`;
    }
    default: {
      throw new Error(`Unrecognized scalar type: ${scalar}`);
    }
  }
}

export function camelCase(str: string) {
  return str[0].toLowerCase() + str.slice(1);
}

export function pascalCase(str: string): string {
  return str[0].toUpperCase() + str.slice(1);
}

function getInputKeywordPhrasePosition(inputTypeName: string) {
  const inputParseResult = [
    "Create",
    "OrderBy",
    "Update",
    "Upsert",
    "ScalarWhere",
    "Where",
    "Filter",
  ]
    .map(inputKeyword => inputTypeName.search(inputKeyword))
    .filter(position => position >= 0);

  if (inputParseResult.length === 0) {
    return;
  }

  const keywordPhrasePosition = inputParseResult[0];
  return keywordPhrasePosition;
}

export function getModelNameFromInputType(inputTypeName: string) {
  const keywordPhrasePosition = getInputKeywordPhrasePosition(inputTypeName);
  if (!keywordPhrasePosition) {
    return;
  }
  const modelName = inputTypeName.slice(0, keywordPhrasePosition);
  return modelName;
}

export function getInputTypeName(
  originalInputName: string,
  dmmfDocument: DmmfDocument,
): string {
  const keywordPhrasePosition = getInputKeywordPhrasePosition(
    originalInputName,
  );
  if (!keywordPhrasePosition) {
    return originalInputName;
  }

  const modelName = originalInputName.slice(0, keywordPhrasePosition);
  const typeNameRest = originalInputName.slice(keywordPhrasePosition);
  const modelTypeName = dmmfDocument.getModelTypeName(modelName);
  if (!modelTypeName) {
    return originalInputName;
  }

  return `${modelTypeName}${typeNameRest}`;
}

export function cleanDocsString(
  documentation: string | undefined,
): string | undefined {
  if (!documentation) {
    return;
  }
  let cleanedDocs = documentation;
  cleanedDocs = cleanedDocs.replace(modelAttributeRegex, "");
  cleanedDocs = cleanedDocs.replace(fieldAttributeRegex, "");
  cleanedDocs = cleanedDocs.split('"').join('\\"');
  cleanedDocs = cleanedDocs.split("\r").join("");
  cleanedDocs = cleanedDocs.split("\\r").join("");
  cleanedDocs = cleanedDocs.split("\n").join("");
  cleanedDocs = cleanedDocs.split("\\n").join("");
  return cleanedDocs;
}

export function toUnixPath(maybeWindowsPath: string) {
  return maybeWindowsPath.split("\\").join("/");
}
