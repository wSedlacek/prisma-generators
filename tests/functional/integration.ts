import "reflect-metadata";
import { promises as fs } from "fs";
import directoryTree from "directory-tree";
import path from "path";
import util from "util";
import childProcess from "child_process";
import { buildSchema } from "type-graphql";
import { graphql } from "graphql";
import pg from "pg";

import generateArtifactsDirPath from "../helpers/artifacts-dir";
import { stringifyDirectoryTrees } from "../helpers/structure";

const exec = util.promisify(childProcess.exec);

describe("generator integration", () => {
  let cwdDirPath: string;
  let schema: string;

  beforeEach(async () => {
    cwdDirPath = generateArtifactsDirPath("functional-integration");
    await fs.mkdir(cwdDirPath, { recursive: true });

    schema = /* prisma */ `
      datasource db {
        provider = "postgresql"
        url      = env("DATABASE_URL")
      }

      generator client {
        provider = "prisma-client-js"
        output   = "./generated/client"
      }

      generator typegraphql {
        provider = "node ../../../src/cli/dev.ts"
        output   = "./generated/type-graphql"
      }

      enum Color {
        RED
        GREEN
        BLUE
      }

      model User {
        id     Int      @id @default(autoincrement())
        name   String?
        posts  Post[]
      }

      model Post {
        uuid      String  @id @default(cuid())
        content   String
        author    User    @relation(fields: [authorId], references: [id])
        authorId  Int
        color     Color
      }
    `;
    await fs.writeFile(path.join(cwdDirPath, "schema.prisma"), schema);
    await fs.writeFile(
      path.join(cwdDirPath, ".env"),
      `DATABASE_URL=${process.env.TEST_DATABASE_URL}`,
    );
  });

  it("should generates TypeGraphQL classes files to output folder by running `prisma generate`", async () => {
    const prismaGenerateResult = await exec("npx prisma generate", {
      cwd: cwdDirPath,
    });
    // console.log(prismaGenerateResult);

    const directoryStructure = directoryTree(
      cwdDirPath + "/generated/type-graphql",
    );
    const directoryStructureString =
      "\n[type-graphql]\n" +
      stringifyDirectoryTrees(directoryStructure.children, 2);

    expect(prismaGenerateResult.stderr).toEqual(
      "\x1B[2mEnvironment variables loaded from current directory\x1B[22m\n",
    );
    expect(directoryStructureString).toMatchSnapshot("files structure");
  }, 60000);

  it("should be able to use generate TypeGraphQL classes files to generate GraphQL schema", async () => {
    const prismaGenerateResult = await exec("npx prisma generate", {
      cwd: cwdDirPath,
    });
    // console.log(prismaGenerateResult);
    const {
      UserCrudResolver,
      PostCrudResolver,
      UserRelationsResolver,
      PostRelationsResolver,
    } = require(cwdDirPath + "/generated/type-graphql");
    await buildSchema({
      resolvers: [
        UserCrudResolver,
        PostCrudResolver,
        UserRelationsResolver,
        PostRelationsResolver,
      ],
      validate: false,
      emitSchemaFile: cwdDirPath + "/schema.graphql",
    });
    const graphQLSchemaSDL = await fs.readFile(cwdDirPath + "/schema.graphql", {
      encoding: "utf8",
    });

    expect(prismaGenerateResult.stderr).toEqual(
      "\x1B[2mEnvironment variables loaded from current directory\x1B[22m\n",
    );
    expect(graphQLSchemaSDL).toMatchSnapshot("graphQLSchemaSDL");
  }, 60000);

  it("should be able to generate TypeGraphQL classes files without any type errors", async () => {
    const tsconfigContent = {
      compilerOptions: {
        target: "es2019",
        module: "commonjs",
        lib: ["es2019"],
        strict: true,
        skipLibCheck: true,
        esModuleInterop: true,
        experimentalDecorators: true,
        emitDecoratorMetadata: true,
        forceConsistentCasingInFileNames: true,
      },
    };
    const typegraphqlfolderPath = path.join(
      cwdDirPath,
      "generated",
      "type-graphql",
    );

    const prismaGenerateResult = await exec("npx prisma generate", {
      cwd: cwdDirPath,
    });
    // console.log(prismaGenerateResult);
    await fs.writeFile(
      path.join(typegraphqlfolderPath, "tsconfig.json"),
      JSON.stringify(tsconfigContent),
    );
    const tscResult = await exec("npx tsc --noEmit", {
      cwd: typegraphqlfolderPath,
    });

    expect(prismaGenerateResult.stderr).toEqual(
      "\x1B[2mEnvironment variables loaded from current directory\x1B[22m\n",
    );
    expect(tscResult.stdout).toHaveLength(0);
    expect(tscResult.stderr).toHaveLength(0);
  }, 60000);

  it("should properly fetch the data from DB using PrismaClient while queried by GraphQL schema", async () => {
    const prismaGenerateResult = await exec("npx prisma generate", {
      cwd: cwdDirPath,
    });
    // console.log(prismaGenerateResult);
    expect(prismaGenerateResult.stderr).toEqual(
      "\x1B[2mEnvironment variables loaded from current directory\x1B[22m\n",
    );

    // drop database before migrate
    const originalDatabaseUrl = process.env.TEST_DATABASE_URL!;
    const [dbName, ...databaseUrlParts] = originalDatabaseUrl
      .split("/")
      .reverse();
    const databaseUrl = databaseUrlParts.reverse().join("/") + "/postgres";
    const pgClient = new pg.Client({
      connectionString: databaseUrl,
    });
    await pgClient.connect();
    await pgClient.query(`DROP DATABASE IF EXISTS "${dbName}"`);
    await pgClient.end();

    const migrateSaveResult = await exec(
      "npx prisma migrate save --experimental --name='init' --create-db",
      { cwd: cwdDirPath },
    );
    // console.log(migrateSaveResult);
    expect(migrateSaveResult.stderr).toEqual(
      "\x1B[2mEnvironment variables loaded from current directory\x1B[22m\n",
    );

    const migrateUpResult = await exec("npx prisma migrate up --experimental", {
      cwd: cwdDirPath,
    });
    // console.log(migrateUpResult);
    expect(migrateUpResult.stderr).toEqual(
      "\x1B[2mEnvironment variables loaded from current directory\x1B[22m\n",
    );

    const { PrismaClient } = require(cwdDirPath + "/generated/client");
    const prisma = new PrismaClient();

    await prisma.user.create({ data: { name: "test1" } });
    await prisma.user.create({
      data: {
        name: "test2",
        posts: {
          create: [
            {
              color: "RED",
              content: "post content",
            },
          ],
        },
      },
    });
    await prisma.user.create({ data: { name: "not test" } });

    const {
      UserCrudResolver,
      PostCrudResolver,
      UserRelationsResolver,
      PostRelationsResolver,
    } = require(cwdDirPath + "/generated/type-graphql");
    const graphQLSchema = await buildSchema({
      resolvers: [
        UserCrudResolver,
        PostCrudResolver,
        UserRelationsResolver,
        PostRelationsResolver,
      ],
      validate: false,
    });

    const query = /* graphql */ `
      query {
        users(where: {
          name: {
            startsWith: "test"
          }
        }) {
          id
          name
          posts {
            content
            color
            author {
              name
            }
          }
        }
      }
    `;
    const { data, errors } = await graphql(graphQLSchema, query, null, {
      prisma,
    });
    await prisma.disconnect();

    expect(errors).toBeUndefined();
    expect(data).toMatchSnapshot("graphql data");
  }, 100000);
});
