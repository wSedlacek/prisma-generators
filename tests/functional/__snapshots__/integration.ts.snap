// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generator integration should be able to use generate TypeGraphQL classes files to generate GraphQL schema: graphQLSchemaSDL 1`] = `
"# -----------------------------------------------
# !!! THIS FILE WAS GENERATED BY TYPE-GRAPHQL !!!
# !!!   DO NOT MODIFY THIS FILE BY YOURSELF   !!!
# -----------------------------------------------

type AggregatePost {
  avg: PostAvgAggregateOutputType
  count: Int!
  max: PostMaxAggregateOutputType
  min: PostMinAggregateOutputType
  sum: PostSumAggregateOutputType
}

type AggregateUser {
  avg: UserAvgAggregateOutputType
  count: Int!
  max: UserMaxAggregateOutputType
  min: UserMinAggregateOutputType
  sum: UserSumAggregateOutputType
}

type BatchPayload {
  count: Int!
}

enum Color {
  BLUE
  GREEN
  RED
}

input ColorFilter {
  equals: Color
  in: [Color!]
  not: Color
  notIn: [Color!]
}

input IntFilter {
  equals: Int
  gt: Int
  gte: Int
  in: [Int!]
  lt: Int
  lte: Int
  not: Int
  notIn: [Int!]
}

type Mutation {
  createPost(data: PostCreateInput!): Post!
  createUser(data: UserCreateInput!): User!
  deleteManyPost(where: PostWhereInput): BatchPayload!
  deleteManyUser(where: UserWhereInput): BatchPayload!
  deletePost(where: PostWhereUniqueInput!): Post
  deleteUser(where: UserWhereUniqueInput!): User
  updateManyPost(data: PostUpdateManyMutationInput!, where: PostWhereInput): BatchPayload!
  updateManyUser(data: UserUpdateManyMutationInput!, where: UserWhereInput): BatchPayload!
  updatePost(data: PostUpdateInput!, where: PostWhereUniqueInput!): Post
  updateUser(data: UserUpdateInput!, where: UserWhereUniqueInput!): User
  upsertPost(create: PostCreateInput!, update: PostUpdateInput!, where: PostWhereUniqueInput!): Post!
  upsertUser(create: UserCreateInput!, update: UserUpdateInput!, where: UserWhereUniqueInput!): User!
}

input NullableStringFilter {
  contains: String
  endsWith: String
  equals: String
  gt: String
  gte: String
  in: [String!]
  lt: String
  lte: String
  not: String
  notIn: [String!]
  startsWith: String
}

enum OrderByArg {
  asc
  desc
}

type Post {
  author: User!
  authorId: Int!
  color: Color!
  content: String!
  uuid: String!
}

type PostAvgAggregateOutputType {
  authorId: Float!
}

input PostCreateInput {
  author: UserCreateOneWithoutPostsInput!
  color: Color!
  content: String!
  uuid: String
}

input PostCreateManyWithoutAuthorInput {
  connect: [PostWhereUniqueInput!]
  create: [PostCreateWithoutAuthorInput!]
}

input PostCreateWithoutAuthorInput {
  color: Color!
  content: String!
  uuid: String
}

input PostFilter {
  every: PostWhereInput
  none: PostWhereInput
  some: PostWhereInput
}

type PostMaxAggregateOutputType {
  authorId: Int!
}

type PostMinAggregateOutputType {
  authorId: Int!
}

input PostOrderByInput {
  authorId: OrderByArg
  color: OrderByArg
  content: OrderByArg
  uuid: OrderByArg
}

input PostScalarWhereInput {
  AND: [PostScalarWhereInput!]
  authorId: IntFilter
  color: ColorFilter
  content: StringFilter
  NOT: [PostScalarWhereInput!]
  OR: [PostScalarWhereInput!]
  uuid: StringFilter
}

type PostSumAggregateOutputType {
  authorId: Int!
}

input PostUpdateInput {
  author: UserUpdateOneRequiredWithoutPostsInput
  color: Color
  content: String
  uuid: String
}

input PostUpdateManyDataInput {
  color: Color
  content: String
  uuid: String
}

input PostUpdateManyMutationInput {
  color: Color
  content: String
  uuid: String
}

input PostUpdateManyWithoutAuthorInput {
  connect: [PostWhereUniqueInput!]
  create: [PostCreateWithoutAuthorInput!]
  delete: [PostWhereUniqueInput!]
  deleteMany: [PostScalarWhereInput!]
  disconnect: [PostWhereUniqueInput!]
  set: [PostWhereUniqueInput!]
  update: [PostUpdateWithWhereUniqueWithoutAuthorInput!]
  updateMany: [PostUpdateManyWithWhereNestedInput!]
  upsert: [PostUpsertWithWhereUniqueWithoutAuthorInput!]
}

input PostUpdateManyWithWhereNestedInput {
  data: PostUpdateManyDataInput!
  where: PostScalarWhereInput!
}

input PostUpdateWithoutAuthorDataInput {
  color: Color
  content: String
  uuid: String
}

input PostUpdateWithWhereUniqueWithoutAuthorInput {
  data: PostUpdateWithoutAuthorDataInput!
  where: PostWhereUniqueInput!
}

input PostUpsertWithWhereUniqueWithoutAuthorInput {
  create: PostCreateWithoutAuthorInput!
  update: PostUpdateWithoutAuthorDataInput!
  where: PostWhereUniqueInput!
}

input PostWhereInput {
  AND: [PostWhereInput!]
  author: UserWhereInput
  authorId: IntFilter
  color: ColorFilter
  content: StringFilter
  NOT: [PostWhereInput!]
  OR: [PostWhereInput!]
  uuid: StringFilter
}

input PostWhereUniqueInput {
  uuid: String
}

type Query {
  aggregatePost(cursor: PostWhereUniqueInput, orderBy: PostOrderByInput, skip: Int, take: Int, where: PostWhereInput): AggregatePost!
  aggregateUser(cursor: UserWhereUniqueInput, orderBy: UserOrderByInput, skip: Int, take: Int, where: UserWhereInput): AggregateUser!
  post(where: PostWhereUniqueInput!): Post
  posts(cursor: PostWhereUniqueInput, orderBy: PostOrderByInput, skip: Int, take: Int, where: PostWhereInput): [Post!]!
  user(where: UserWhereUniqueInput!): User
  users(cursor: UserWhereUniqueInput, orderBy: UserOrderByInput, skip: Int, take: Int, where: UserWhereInput): [User!]!
}

input StringFilter {
  contains: String
  endsWith: String
  equals: String
  gt: String
  gte: String
  in: [String!]
  lt: String
  lte: String
  not: String
  notIn: [String!]
  startsWith: String
}

type User {
  id: Int!
  name: String
  posts(cursor: PostWhereUniqueInput, orderBy: PostOrderByInput, skip: Int, take: Int, where: PostWhereInput): [Post!]
}

type UserAvgAggregateOutputType {
  id: Float!
}

input UserCreateInput {
  name: String
  posts: PostCreateManyWithoutAuthorInput
}

input UserCreateOneWithoutPostsInput {
  connect: UserWhereUniqueInput
  create: UserCreateWithoutPostsInput
}

input UserCreateWithoutPostsInput {
  name: String
}

type UserMaxAggregateOutputType {
  id: Int!
}

type UserMinAggregateOutputType {
  id: Int!
}

input UserOrderByInput {
  id: OrderByArg
  name: OrderByArg
}

type UserSumAggregateOutputType {
  id: Int!
}

input UserUpdateInput {
  name: String
  posts: PostUpdateManyWithoutAuthorInput
}

input UserUpdateManyMutationInput {
  name: String
}

input UserUpdateOneRequiredWithoutPostsInput {
  connect: UserWhereUniqueInput
  create: UserCreateWithoutPostsInput
  update: UserUpdateWithoutPostsDataInput
  upsert: UserUpsertWithoutPostsInput
}

input UserUpdateWithoutPostsDataInput {
  name: String
}

input UserUpsertWithoutPostsInput {
  create: UserCreateWithoutPostsInput!
  update: UserUpdateWithoutPostsDataInput!
}

input UserWhereInput {
  AND: [UserWhereInput!]
  id: IntFilter
  name: NullableStringFilter
  NOT: [UserWhereInput!]
  OR: [UserWhereInput!]
  posts: PostFilter
}

input UserWhereUniqueInput {
  id: Int
}
"
`;

exports[`generator integration should generates TypeGraphQL classes files to output folder by running \`prisma generate\`: files structure 1`] = `
"
[type-graphql]
  [enums]
    Color.ts
    OrderByArg.ts
    index.ts
  index.ts
  [models]
    Post.ts
    User.ts
    index.ts
  [resolvers]
    [crud]
      [Post]
        AggregatePostResolver.ts
        CreatePostResolver.ts
        DeleteManyPostResolver.ts
        DeletePostResolver.ts
        FindManyPostResolver.ts
        FindOnePostResolver.ts
        PostCrudResolver.ts
        UpdateManyPostResolver.ts
        UpdatePostResolver.ts
        UpsertPostResolver.ts
        [args]
          AggregatePostArgs.ts
          CreatePostArgs.ts
          DeleteManyPostArgs.ts
          DeletePostArgs.ts
          FindManyPostArgs.ts
          FindOnePostArgs.ts
          UpdateManyPostArgs.ts
          UpdatePostArgs.ts
          UpsertPostArgs.ts
          index.ts
      [User]
        AggregateUserResolver.ts
        CreateUserResolver.ts
        DeleteManyUserResolver.ts
        DeleteUserResolver.ts
        FindManyUserResolver.ts
        FindOneUserResolver.ts
        UpdateManyUserResolver.ts
        UpdateUserResolver.ts
        UpsertUserResolver.ts
        UserCrudResolver.ts
        [args]
          AggregateUserArgs.ts
          CreateUserArgs.ts
          DeleteManyUserArgs.ts
          DeleteUserArgs.ts
          FindManyUserArgs.ts
          FindOneUserArgs.ts
          UpdateManyUserArgs.ts
          UpdateUserArgs.ts
          UpsertUserArgs.ts
          index.ts
      index.ts
    [inputs]
      ColorFilter.ts
      IntFilter.ts
      NullableStringFilter.ts
      PostCreateInput.ts
      PostCreateManyWithoutAuthorInput.ts
      PostCreateWithoutAuthorInput.ts
      PostFilter.ts
      PostOrderByInput.ts
      PostScalarWhereInput.ts
      PostUpdateInput.ts
      PostUpdateManyDataInput.ts
      PostUpdateManyMutationInput.ts
      PostUpdateManyWithWhereNestedInput.ts
      PostUpdateManyWithoutAuthorInput.ts
      PostUpdateWithWhereUniqueWithoutAuthorInput.ts
      PostUpdateWithoutAuthorDataInput.ts
      PostUpsertWithWhereUniqueWithoutAuthorInput.ts
      PostWhereInput.ts
      PostWhereUniqueInput.ts
      StringFilter.ts
      UserCreateInput.ts
      UserCreateOneWithoutPostsInput.ts
      UserCreateWithoutPostsInput.ts
      UserOrderByInput.ts
      UserUpdateInput.ts
      UserUpdateManyMutationInput.ts
      UserUpdateOneRequiredWithoutPostsInput.ts
      UserUpdateWithoutPostsDataInput.ts
      UserUpsertWithoutPostsInput.ts
      UserWhereInput.ts
      UserWhereUniqueInput.ts
      index.ts
    [outputs]
      AggregatePost.ts
      AggregateUser.ts
      BatchPayload.ts
      PostAvgAggregateOutputType.ts
      PostMaxAggregateOutputType.ts
      PostMinAggregateOutputType.ts
      PostSumAggregateOutputType.ts
      UserAvgAggregateOutputType.ts
      UserMaxAggregateOutputType.ts
      UserMinAggregateOutputType.ts
      UserSumAggregateOutputType.ts
      index.ts
    [relations]
      [Post]
        PostRelationsResolver.ts
      [User]
        UserRelationsResolver.ts
        [args]
          UserPostsArgs.ts
          index.ts
      index.ts
"
`;

exports[`generator integration should properly fetch the data from DB using PrismaClient while queried by GraphQL schema: graphql data 1`] = `
Object {
  "users": Array [
    Object {
      "id": 1,
      "name": "test1",
      "posts": Array [],
    },
    Object {
      "id": 2,
      "name": "test2",
      "posts": Array [
        Object {
          "author": Object {
            "name": "test2",
          },
          "color": "RED",
          "content": "post content",
        },
      ],
    },
  ],
}
`;
